src/Compilers/VisualBasic/Portable/SourceGeneration/VisualBasicGeneratorDriver.vb

```diff
@@ -32,7 +34,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
         Friend Overrides Function ParseGeneratedSourceText(input As GeneratedSourceText, fileName As String, cancellationToken As CancellationToken) As SyntaxTree
-            Return SyntaxFactory.ParseSyntaxTree(input.Text, _state.ParseOptions, fileName, cancellationToken)
+            Dim existingTree As SyntaxTree = Nothing
+            If s_parsedGeneratedSources.TryGetValue(input.Text, existingTree) _
+                AndAlso Equals(_state.ParseOptions, existingTree.Options) _
+                AndAlso Equals(fileName, existingTree.FilePath) Then
+                Return existingTree
+            End If
+
+            Dim tree = SyntaxFactory.ParseSyntaxTree(input.Text, _state.ParseOptions, fileName, cancellationToken)
+
+#If NETCOREAPP Then
+            s_parsedGeneratedSources.AddOrUpdate(input.Text, tree)
+#Else
+            s_parsedGeneratedSources.Remove(input.Text)
+            s_parsedGeneratedSources.Add(input.Text, tree)
+#End If
+
+            Return tree
```

Comment:
minor nit.  but it seems like we could have an abstract base class for these that has the logic once.

Reply:
I moved the logic to the base class.